### PR TITLE
✨ internalize block proposer logic

### PR DESCRIPTION
**Pull request by abdelhamidbakhta**
Extract the code from [subtrate basic authorship](https://github.com/paritytech/substrate/tree/master/client/basic-authorship) and integrate it in Madara client.

Changes summary:
- create new crate `mc-block-proposer` in `crates/client/block-proposer`
- adapt `Proposer` implementation logic
    - remove hashing of extrinsics to log information on the proposed block (performance optimization and less verbose logging)
    - remove unneeded telemetry reports
    - remove `include_proof_in_estimation` field

### CODE REVIEW

File: crates/node/Cargo.toml [hunk 1]
```diff
@@ -72,11 +71,13 @@ frame-benchmarking = { workspace = true }
 frame-benchmarking-cli = { workspace = true }
 
 # Starknet
+# FIXME: declare those dependencies on workspace level
```
Comment (LucasLvy) [inline crates/node/Cargo.toml:74]: add issue nb

**Review by LucasLvy — CHANGES_REQUESTED** [review]

File: crates/client/block-proposer/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+/// Default block size limit in bytes used by [`Proposer`].
+///
+/// Can be overwritten by [`ProposerFactory::set_default_block_size_limit`].
+///
+/// Be aware that there is also an upper packet size on what the networking code
+/// will accept. If the block doesn't fit in such a package, it can not be
+/// transferred to other nodes.
+pub const DEFAULT_BLOCK_SIZE_LIMIT: usize = 4 * 1024 * 1024 + 512;
+
+const DEFAULT_SOFT_DEADLINE_PERCENT: Percent = Percent::from_percent(80);
```
Comment (LucasLvy) [inline crates/client/block-proposer/src/lib.rs:39]: what is that ?

**Review by LucasLvy — CHANGES_REQUESTED** [review]

File: crates/client/block-proposer/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+        info!(
+            "🥷 Prepared block for proposing at {} ({} ms) [hash: {:?}; parent_hash: {}; extrinsics ({})]",
+            block.header().number(),
+            block_timer.elapsed().as_millis(),
+            block.header().hash(),
+            block.header().parent_hash(),
+            block.extrinsics().len(),
+        );
+
+        // TODO: try to remove completely the storage proof computation and see if it breaks anything
```
Comment (LucasLvy) [inline crates/client/block-proposer/src/lib.rs:397]: issue ?

File: crates/node/Cargo.toml [hunk 4]
```diff
@@ -72,11 +71,13 @@ frame-benchmarking = { workspace = true }
...
 madara-runtime = { path = "../runtime" }
 mc-db = { path = "../client/db" }
 mc-rpc = { path = "../client/rpc" }
 mc-storage = { path = "../client/storage" }
 mc-mapping-sync = { path = "../client/mapping-sync" }
+mc-block-proposer = {workspace = true}
```
Comment (LucasLvy) [inline crates/node/Cargo.toml:80]: format that

**Review by LucasLvy — CHANGES_REQUESTED** [review]

File: crates/client/block-proposer/src/lib.rs (hunk 2 above)
Comment (LucasLvy) [reply crates/client/block-proposer/src/lib.rs:39]: so no comment ?

File: crates/client/block-proposer/src/lib.rs (hunk 3 above)
Comment (abdelhamidbakhta) [reply crates/client/block-proposer/src/lib.rs:397]: done

**Review by LucasLvy — APPROVED** [review]

**Pull request merged**